[quant][graphmode] Add observers for dynamic quant

supriyar opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #35265 [quant][graph] Add a new observer type for dynamic quantization
* #35235 [quant][graph] Add _choose_qparams function for graph mode
* **#35121 [quant][graphmode] Add observers for dynamic quant**

Summary:
For dynamic quantization we insert observers at the input to mimic the quatization of activations that happens in the operator
Observer for weight is inserted similar to static quant

Test Plan:
python test/test_quantize_script.py

Sample output for single layer FC

.graph(%self : __torch__.___torch_mangle_4.M,
      %x.2 : Tensor):
  %_observer_1 : __torch__.torch.quantization.observer.MinMaxObserver = prim::GetAttr[name="_observer_1"](%self)
  %x.1 : Tensor = prim::CallMethod[name="forward"](%_observer_1, %x.2)
  %2 : __torch__.torch.nn.modules.linear.___torch_mangle_5.Linear = prim::GetAttr[name="fc"](%self)
  %3 : Tensor = prim::CallMethod[name="forward"](%2, %x.1) # test/test_quantize_script.py:19:23
  return (%3)

graph(%self : __torch__.torch.nn.modules.linear.___torch_mangle_5.Linear,
      %input.1 : Tensor):
 %2 : Function = prim::Constant[name="linear"]()
 %3 : Tensor = prim::GetAttr[name="weight"](%self)
 %_observer_0 : __torch__.torch.quantization.observer.MinMaxObserver = prim::GetAttr[name="_observer_0"](%self)
 %7 : Tensor = prim::CallMethod[name="forward"](%_observer_0, %3)
 %4 : Tensor = prim::GetAttr[name="bias"](%self)
 %5 : Tensor = prim::CallFunction(%2, %input.1, %7, %4) # /home/supriyar/miniconda3/envs/pytorch_py3/lib/python3.7/site-packages/torch/nn/modules/linear.py:87:15
 return (%5)

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20599144](https://our.internmc.facebook.com/intern/diff/D20599144)

supriyar committed:
[quant][graphmode] Add observers for dynamic quant
Summary:
For dynamic quantization we insert observers at the input to mimic the quatization of activations that happens in the operator
Observer for weight is inserted similar to static quant

Test Plan:
python test/test_quantize_script.py

Sample output for single layer FC

.graph(%self : __torch__.___torch_mangle_4.M,
      %x.2 : Tensor):
  %_observer_1 : __torch__.torch.quantization.observer.MinMaxObserver = prim::GetAttr[name="_observer_1"](%self)
  %x.1 : Tensor = prim::CallMethod[name="forward"](%_observer_1, %x.2)
  %2 : __torch__.torch.nn.modules.linear.___torch_mangle_5.Linear = prim::GetAttr[name="fc"](%self)
  %3 : Tensor = prim::CallMethod[name="forward"](%2, %x.1) # test/test_quantize_script.py:19:23
  return (%3)

graph(%self : __torch__.torch.nn.modules.linear.___torch_mangle_5.Linear,
      %input.1 : Tensor):
 %2 : Function = prim::Constant[name="linear"]()
 %3 : Tensor = prim::GetAttr[name="weight"](%self)
 %_observer_0 : __torch__.torch.quantization.observer.MinMaxObserver = prim::GetAttr[name="_observer_0"](%self)
 %7 : Tensor = prim::CallMethod[name="forward"](%_observer_0, %3)
 %4 : Tensor = prim::GetAttr[name="bias"](%self)
 %5 : Tensor = prim::CallFunction(%2, %input.1, %7, %4) # /home/supriyar/miniconda3/envs/pytorch_py3/lib/python3.7/site-packages/torch/nn/modules/linear.py:87:15
 return (%5)

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar commented on test/test_quantize_script.py:
```diff
@@ -0,0 +1,68 @@
+        assert len(attrs_with_prefix(m.sub, '_observer_')) == 0
+        # weight of linear
+        assert len(attrs_with_prefix(m.sub.fc, '_observer_')) == 1
+
```
@jerryzh168 let me know if you think anything else needs to be tested here.

jerryzh168 replied:
I think it might be better to also test for the sequence of operations, to check the observer is for input, not for output, you can use something like: https://github.com/pytorch/pytorch/blob/master/test/test_jit.py#L1217-L1223

supriyar committed:
Update on "[quant][graphmode] Add observers for dynamic quant"


Summary:
For dynamic quantization we insert observers at the input to mimic the quatization of activations that happens in the operator
Observer for weight is inserted similar to static quant

Test Plan:
python test/test_quantize_script.py

Sample output for single layer FC

.graph(%self : __torch__.___torch_mangle_4.M,
      %x.2 : Tensor):
  %_observer_1 : __torch__.torch.quantization.observer.MinMaxObserver = prim::GetAttr[name="_observer_1"](%self)
  %x.1 : Tensor = prim::CallMethod[name="forward"](%_observer_1, %x.2)
  %2 : __torch__.torch.nn.modules.linear.___torch_mangle_5.Linear = prim::GetAttr[name="fc"](%self)
  %3 : Tensor = prim::CallMethod[name="forward"](%2, %x.1) # test/test_quantize_script.py:19:23
  return (%3)

graph(%self : __torch__.torch.nn.modules.linear.___torch_mangle_5.Linear,
      %input.1 : Tensor):
 %2 : Function = prim::Constant[name="linear"]()
 %3 : Tensor = prim::GetAttr[name="weight"](%self)
 %_observer_0 : __torch__.torch.quantization.observer.MinMaxObserver = prim::GetAttr[name="_observer_0"](%self)
 %7 : Tensor = prim::CallMethod[name="forward"](%_observer_0, %3)
 %4 : Tensor = prim::GetAttr[name="bias"](%self)
 %5 : Tensor = prim::CallFunction(%2, %input.1, %7, %4) # /home/supriyar/miniconda3/envs/pytorch_py3/lib/python3.7/site-packages/torch/nn/modules/linear.py:87:15
 return (%5)

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented on test/test_quantize_script.py:
```diff
@@ -0,0 +1,68 @@
+        qconfig_dict = {
+            'sub.fc': qconfig
+        }
+        m = wrap_cpp_module(torch._C._jit_pass_insert_observers(m._c, "forward",
```
please add a `prepare_dynamic_script` for dynamic quant as well, in `_quantize_script.py`, also here we should call that API I think.

supriyar replied:
Thought of doing it later, but makes sense to add that in the same PR.

jerryzh168 commented on test/test_quantize_script.py:
```diff
@@ -0,0 +1,68 @@
+        # for weight
+        assert len(attrs_with_prefix(m.fc, '_observer_')) == 1
+
+    def test_insert_observers_child_dynamic_qconfig(self):
```
nit: this should be test_insert_observers_dynamic_child_qconfig

jerryzh168 commented on test/test_quantize_script.py:
```diff
@@ -0,0 +1,68 @@
+
+        m = torch.jit.script(M())
+        observer = torch.jit.script(default_observer())
+        qconfig_dict = {'': script_qconfig(default_qconfig)}
```
let's use `default_qconfig` directly

jerryzh168 commented on test/test_quantize_script.py:
```diff
@@ -0,0 +1,68 @@
+from torch.testing._internal.jit_utils import JitTestCase
+
+class TestScript(JitTestCase):
+    def test_insert_observers_dynamic(self):
```
let's use `prepare_dynamic` instead of `insert_observers_dynamic` since that is the API we are going to expose to user

supriyar committed:
Update on "[quant][graphmode] Add observers for dynamic quant"


Summary:
For dynamic quantization we insert observers at the input to mimic the quatization of activations that happens in the operator
Observer for weight is inserted similar to static quant

Test Plan:
python test/test_quantize_script.py

Sample output for single layer FC

.graph(%self : __torch__.___torch_mangle_4.M,
      %x.2 : Tensor):
  %_observer_1 : __torch__.torch.quantization.observer.MinMaxObserver = prim::GetAttr[name="_observer_1"](%self)
  %x.1 : Tensor = prim::CallMethod[name="forward"](%_observer_1, %x.2)
  %2 : __torch__.torch.nn.modules.linear.___torch_mangle_5.Linear = prim::GetAttr[name="fc"](%self)
  %3 : Tensor = prim::CallMethod[name="forward"](%2, %x.1) # test/test_quantize_script.py:19:23
  return (%3)

graph(%self : __torch__.torch.nn.modules.linear.___torch_mangle_5.Linear,
      %input.1 : Tensor):
 %2 : Function = prim::Constant[name="linear"]()
 %3 : Tensor = prim::GetAttr[name="weight"](%self)
 %_observer_0 : __torch__.torch.quantization.observer.MinMaxObserver = prim::GetAttr[name="_observer_0"](%self)
 %7 : Tensor = prim::CallMethod[name="forward"](%_observer_0, %3)
 %4 : Tensor = prim::GetAttr[name="bias"](%self)
 %5 : Tensor = prim::CallFunction(%2, %input.1, %7, %4) # /home/supriyar/miniconda3/envs/pytorch_py3/lib/python3.7/site-packages/torch/nn/modules/linear.py:87:15
 return (%5)

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -446,7 +446,8 @@ class InsertObserversHelper {
           std::unordered_set<Value*>());
-
+  // Is dynamic quantization enabled for the observer pass.
+  bool is_dynamic = false;
```
maybe make this private and expose this through a function like `set_dynamic_flag`

supriyar committed:
Update on "[quant][graphmode] Add observers for dynamic quant"


Summary:
For dynamic quantization we insert observers at the input to mimic the quatization of activations that happens in the operator
Observer for weight is inserted similar to static quant

Test Plan:
python test/test_quantize_script.py

Sample output for single layer FC

.graph(%self : __torch__.___torch_mangle_4.M,
      %x.2 : Tensor):
  %_observer_1 : __torch__.torch.quantization.observer.MinMaxObserver = prim::GetAttr[name="_observer_1"](%self)
  %x.1 : Tensor = prim::CallMethod[name="forward"](%_observer_1, %x.2)
  %2 : __torch__.torch.nn.modules.linear.___torch_mangle_5.Linear = prim::GetAttr[name="fc"](%self)
  %3 : Tensor = prim::CallMethod[name="forward"](%2, %x.1) # test/test_quantize_script.py:19:23
  return (%3)

graph(%self : __torch__.torch.nn.modules.linear.___torch_mangle_5.Linear,
      %input.1 : Tensor):
 %2 : Function = prim::Constant[name="linear"]()
 %3 : Tensor = prim::GetAttr[name="weight"](%self)
 %_observer_0 : __torch__.torch.quantization.observer.MinMaxObserver = prim::GetAttr[name="_observer_0"](%self)
 %7 : Tensor = prim::CallMethod[name="forward"](%_observer_0, %3)
 %4 : Tensor = prim::GetAttr[name="bias"](%self)
 %5 : Tensor = prim::CallFunction(%2, %input.1, %7, %4) # /home/supriyar/miniconda3/envs/pytorch_py3/lib/python3.7/site-packages/torch/nn/modules/linear.py:87:15
 return (%5)

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -805,6 +808,10 @@ void InsertObserversHelper::fillBoundaryValueMap(
   }
 }
 
+void InsertObserversHelper::set_dynamic_flag(bool is_dynamic_) {
```
sorry, should be `setDynamicFlag`....

jerryzh168 commented on torch/quantization/_quantize_script.py:
```diff
@@ -76,6 +76,17 @@ def prepare_script(model, qconfig_dict, inplace=False):
 
+def prepare_dynamic_script(model, qconfig_dict, inplace=False):
+    _check_is_script_module(model)
+    if not inplace:
```
also we don't want to have inplace flag, you can remove it now

jerryzh168 replied:
we'll have a separate PR to remove all inplace flag for insert observers, I haven't got time to do that

supriyar replied:
Will you be removing it from the pass API as well?

jerryzh168 replied:
yes

supriyar replied:
So is inplace true or false by default?

jerryzh168 replied:
I mean we'll revisit after the `clone` API is ready, current inplace doesn't really make sense and should be removed

jerryzh168 replied:
`inplace` should be False

supriyar replied:
qq why do we have `model = model.copy()` here? seems redundant.

jerryzh168 commented:
This is some legacy code from long time ago

Sent from my iPhone

On Mar 20, 2020, at 18:49, supriyar <notifications@github.com> wrote:

﻿

@supriyar commented on this pull request.

________________________________

In torch/quantization/_quantize_script.py<https://github.com/pytorch/pytorch/pull/35121#discussion_r395949993>:

> @@ -76,6 +76,17 @@ def prepare_script(model, qconfig_dict, inplace=False):
                                                                 False))
     return model

+def prepare_dynamic_script(model, qconfig_dict, inplace=False):
+    _check_is_script_module(model)
+    if not inplace:


qq why do we have model = model.copy() here? seems redundant.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/pytorch/pytorch/pull/35121#discussion_r395949993>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/ABF2R2NXCCWCASR6N6XZUC3RIQMI3ANCNFSM4LQUVSZQ>.

supriyar committed:
Update on "[quant][graphmode] Add observers for dynamic quant"


Summary:
For dynamic quantization we insert observers at the input to mimic the quatization of activations that happens in the operator
Observer for weight is inserted similar to static quant

Test Plan:
python test/test_quantize_script.py

Sample output for single layer FC

.graph(%self : __torch__.___torch_mangle_4.M,
      %x.2 : Tensor):
  %_observer_1 : __torch__.torch.quantization.observer.MinMaxObserver = prim::GetAttr[name="_observer_1"](%self)
  %x.1 : Tensor = prim::CallMethod[name="forward"](%_observer_1, %x.2)
  %2 : __torch__.torch.nn.modules.linear.___torch_mangle_5.Linear = prim::GetAttr[name="fc"](%self)
  %3 : Tensor = prim::CallMethod[name="forward"](%2, %x.1) # test/test_quantize_script.py:19:23
  return (%3)

graph(%self : __torch__.torch.nn.modules.linear.___torch_mangle_5.Linear,
      %input.1 : Tensor):
 %2 : Function = prim::Constant[name="linear"]()
 %3 : Tensor = prim::GetAttr[name="weight"](%self)
 %_observer_0 : __torch__.torch.quantization.observer.MinMaxObserver = prim::GetAttr[name="_observer_0"](%self)
 %7 : Tensor = prim::CallMethod[name="forward"](%_observer_0, %3)
 %4 : Tensor = prim::GetAttr[name="bias"](%self)
 %5 : Tensor = prim::CallFunction(%2, %input.1, %7, %4) # /home/supriyar/miniconda3/envs/pytorch_py3/lib/python3.7/site-packages/torch/nn/modules/linear.py:87:15
 return (%5)

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 reviewed (approved):
LG, thanks

supriyar committed:
Update on "[quant][graphmode] Add observers for dynamic quant"


Summary:
For dynamic quantization we insert observers at the input to mimic the quatization of activations that happens in the operator
Observer for weight is inserted similar to static quant

Test Plan:
python test/test_quantize_script.py

Sample output for single layer FC

.graph(%self : __torch__.___torch_mangle_4.M,
      %x.2 : Tensor):
  %_observer_1 : __torch__.torch.quantization.observer.MinMaxObserver = prim::GetAttr[name="_observer_1"](%self)
  %x.1 : Tensor = prim::CallMethod[name="forward"](%_observer_1, %x.2)
  %2 : __torch__.torch.nn.modules.linear.___torch_mangle_5.Linear = prim::GetAttr[name="fc"](%self)
  %3 : Tensor = prim::CallMethod[name="forward"](%2, %x.1) # test/test_quantize_script.py:19:23
  return (%3)

graph(%self : __torch__.torch.nn.modules.linear.___torch_mangle_5.Linear,
      %input.1 : Tensor):
 %2 : Function = prim::Constant[name="linear"]()
 %3 : Tensor = prim::GetAttr[name="weight"](%self)
 %_observer_0 : __torch__.torch.quantization.observer.MinMaxObserver = prim::GetAttr[name="_observer_0"](%self)
 %7 : Tensor = prim::CallMethod[name="forward"](%_observer_0, %3)
 %4 : Tensor = prim::GetAttr[name="bias"](%self)
 %5 : Tensor = prim::CallFunction(%2, %input.1, %7, %4) # /home/supriyar/miniconda3/envs/pytorch_py3/lib/python3.7/site-packages/torch/nn/modules/linear.py:87:15
 return (%5)

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20599144](https://our.internmc.facebook.com/intern/diff/D20599144)

[ghstack-poisoned]
